staging/src/k8s.io/client-go/generic/type.go

```diff
@@ -0,0 +1,360 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generic
+
+import (
+	"context"
+	json "encoding/json"
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	types "k8s.io/apimachinery/pkg/types"
+	watch "k8s.io/apimachinery/pkg/watch"
+	rest "k8s.io/client-go/rest"
+)
+
+// objectWithMeta matches objects implementing both runtime.Object and metav1.Object.
+type objectWithMeta interface {
+	runtime.Object
+	metav1.Object
+}
+
+// namedObject matches comparable objects implementing GetName(); it is intended for use with apply declarative configurations.
+type namedObject interface {
+	comparable
+	GetName() *string
+}
+
+// TypeClient represents a client, optionally namespaced, with no support for lists or apply declarative configurations.
+type TypeClient[T objectWithMeta] struct {
+	Resource       string
+	Client         rest.Interface
+	Namespace      string
+	newObject      func() T
+	parameterCodec runtime.ParameterCodec
+}
+
+// TypeClientWithList represents a client with support for lists.
+type TypeClientWithList[T objectWithMeta, L runtime.Object] struct {
+	TypeClient[T]
+	newList func() L
+}
+
+// TypeClientWithApply represents a client with support for apply declarative configurations.
+type TypeClientWithApply[T objectWithMeta, C namedObject] struct {
+	TypeClient[T]
+}
+
+// TypeClientWithListAndApply represents a client with support for lists and apply declarative configurations.
+// It can't use both TypeClientWithList and TypeClientWithApply, because of the resulting ambiguous inclusion of TypeClient.
+// Here it uses TypeClientWithApply and adds newList(); this means that list-related methods must be defined twice,
+// once for TypeClientWithList and once for TypeClientWithListAndApply.
+type TypeClientWithListAndApply[T objectWithMeta, L runtime.Object, C namedObject] struct {
```

Comment:
I like the idea, I’ve updated the code to do this (as a separate commit for review purposes). It does end up being a bit weird because the `alsoApplier` and `alsoLister` methods need access to the “surrounding” `TypeClient`, I can’t think of a better way to handle this right now.